code/modules/antagonists/heretic/magic/cosmic_runes.dm

```diff
@@ -0,0 +1,136 @@
+/datum/action/cooldown/spell/cosmic_rune
+	name = "Cosmic Rune"
+	desc = "Creates a cosmic rune at your position, only two can exist at a time. Invoking one rune transports you to the other."
+	background_icon_state = "bg_heretic"
+	overlay_icon_state = "bg_heretic_border"
+	button_icon = 'icons/mob/actions/actions_ecult.dmi'
+	button_icon_state = "cosmic_rune"
+
+	sound = 'sound/magic/forcewall.ogg'
+	school = SCHOOL_FORBIDDEN
+	cooldown_time = 15 SECONDS
+
+	invocation = "ST'R R'N'"
+	invocation_type = INVOCATION_WHISPER
+	spell_requirements = NONE
+
+	/// Storage for the first rune.
+	var/datum/weakref/first_rune
+	/// Storage for the second rune.
+	var/datum/weakref/second_rune
+	/// Rune removal effect.
+	var/obj/effect/rune_remove_effect = /obj/effect/temp_visual/cosmic_rune_fade
+
+/datum/action/cooldown/spell/cosmic_rune/cast(atom/cast_on)
+	. = ..()
+	var/obj/effect/cosmic_rune/first_rune_resolved = first_rune?.resolve()
+	var/obj/effect/cosmic_rune/second_rune_resolved = second_rune?.resolve()
+	if(first_rune_resolved && second_rune_resolved)
+		var/obj/effect/cosmic_rune/new_rune = new /obj/effect/cosmic_rune(get_turf(cast_on))
+		new rune_remove_effect(get_turf(first_rune_resolved))
+		QDEL_NULL(first_rune_resolved)
+		first_rune = WEAKREF(second_rune_resolved)
+		second_rune = WEAKREF(new_rune)
+		second_rune_resolved.link_rune(new_rune)
+		new_rune.link_rune(second_rune_resolved)
+		return
+	if(!first_rune_resolved)
+		var/obj/effect/cosmic_rune/new_rune = new /obj/effect/cosmic_rune(get_turf(cast_on))
+		first_rune = WEAKREF(new_rune)
+		if(second_rune_resolved)
+			new_rune.link_rune(second_rune_resolved)
+			second_rune_resolved.link_rune(new_rune)
+		return
+	if(!second_rune_resolved)
+		var/obj/effect/cosmic_rune/new_rune = new /obj/effect/cosmic_rune(get_turf(cast_on))
+		second_rune = WEAKREF(new_rune)
+		if(first_rune_resolved)
+			first_rune_resolved.link_rune(new_rune)
+			new_rune.link_rune(first_rune_resolved)
```

Comment:
```suggestion
	if(!first_rune_resolved)
		first_rune = make_new_rune((get_turf(cast_on), second_rune_resolved)
		return
	if(!second_rune_resolved)
		second_rune = make_new_rune((get_turf(cast_on), first_rune_resolved)

/// Returns a weak reference to a new rune, linked to an existing rune if provided
/datum/action/cooldown/spell/cosmic_rune/proc/make_new_rune(turf/target_turf, obj/effect/cosmic_rune/other_rune)
	 var/obj/effect/cosmic_rune/new_rune = new /obj/effect/cosmic_rune(target_turf)
	 if (other_rune)
	 	other_rune.link_rune(new_rune)
	 	new_rune.link_rune(other_rune)
	 
	 return WEAKREF(new_rune)
```

Now that I think about it these are exactly the same operation, procs aren't "free" but this isn't hot code